rfcs/0003-timestamps-and-ttl.md

```diff
@@ -0,0 +1,295 @@
+   deleted data is collected within a time bound is left for a future exercise.
+2. **Resolving Out-of-Order Operations**: some applications may process incoming requests
+   in a manner that results in out-of-order insertions to SlateDB (stream processing applications
+   are frequently subject to this type of pattern). Using insertion timestamps to resolve which
+   version of a key is more recent may be preferable to using the clock at the time of insertion.
+3. **Historical Reads / MVCC**: while SlateDB provides snapshot capabilities to ensure reads are
+   consistent and repeatable, it requires a snapshot to have been taken at the desired time. 
+   increasing the number of snapshots adds significant storage overhead. Instead, it is possible
+   to trade read-latency for space amplification by maintaining multiple versions of keys at
+   different timestamps to allow for similar "historical" read operations.
+
+## Goals
+
+- Define the API for specifying timestamps
+- Define the changes to the on-disk format of stored rows
+- Define the default semantics (and enumerate configurations) for out-of-order insertions
+- Outline mechanism to support Time-To-Live (TTL) via Compaction Filters
+
+## Out of Scope
+
+While the following are out of scope for the initial implementation, the design should
+not preclude:
+
+- Outlining compaction strategies that are optimized for TTL (such as Cassandra's
+  [TWCS](https://cassandra.apache.org/doc/stable/cassandra/operating/compaction/twcs.html)). This
+  will require additional metadata per-SST and a compaction strategy to match, but is otherwise
+  possible with what's outlined in the scope of this design.
+- Outlining compaction strategies that target efficient deletes (such as outlined
+  in [Lethe](https://disc-projects.bu.edu/lethe/))
+- Ensuring deletions are resolved within a time bound (a TTL does not ensure the data is deleted
+  from the object store within the specified time)
+- A public API for compaction filters (see https://github.com/slatedb/slatedb/issues/225) - this RFC
+  will implement TTL using a compaction filter but will not yet expose a public API for plugging in
+  custom compaction filters. Instead, the implementation should ensure that such an extension is
+  easy.
+
+## Public API
+
+### Configuration
+
+We will make the following changes to `DbOptions`:
+- introduce a configuration for a custom clock, and will default to using the system clock
+- introduce a configuration for a default TTL, which defaults to -1
+
+```rust
+/// defines the clock that SlateDB will use during this session
+pub trait Clock {
+    fn now(&self) -> i64;
+}
+
+/// contains the default implementation of the Clock, and will return the system time
+pub struct SystemClock;
+
+// ...
+
+pub struct DbOptions {
+    // ...
+
+    /// the Clock to use for insertion timestamps
+    ///
+    /// default: the default clock uses the local system time on the machine
+    clock: Option<Arc<dyn Clock>>,
```

Comment:
Do we have to use dyn Clock? Could we use `DbOptions<T: Clock>` and `clock: Option<Arc<T>>` instead of `dyn Clock`?

Reply:
Sure, I'm not very proficient with Rust -- do you mind explaining the tradeoffs here?

Reply:
Why is that preferable @KaguraMilet? We'd have to put that type parameter in all the places that `DbOptions` gets used/stored, which is kind of annoying.

Reply:
My thinking is simple, using `dyn` introduces unnecessary vtable and runtime overheads. I think we can do this perfectly well with trait bound + generic.

Reply:
[Rust trait object reference](https://doc.rust-lang.org/reference/types/trait-object.html)

> The purpose of trait objects is to permit “late binding” of methods. Calling a method on a trait object results in virtual dispatch at runtime: that is, a function pointer is loaded from the trait object vtable and invoked indirectly. The actual implementation for each vtable entry can vary on an object-by-object basis.

I took a snippet from the Rust reference

Reply:
IMO if it makes the code more verbose I think this may be a premature optimization. the clock is only called once per `put` (as opposed to every document retrieved) so the overhead of a vtable lookup is negligible.

Reply:
+1 For config, I think it's worth making it readable not optimized.

Reply:
i also prefer using `dyn trait` when it's considered not in the hot path in my code , type bounds are still scary for me, it tends to infect all the code depends on it.